File: Delphi-source-code/Providers/ProvidersSupplies/src/uProviderSuppliesBL.pas
unit uProviderSuppliesBL;

interface

uses
  System.SysUtils;

type
  TSupplyItem = record
    SupplyId: Integer;
    ProviderId: Integer;
    ProviderName: string;
    ItemName: string;
    Category: string;
    UnitCost: Currency;
    MinOrderQty: Integer;
    LeadTimeDays: Integer;
    InStock: Boolean;
  end;

  TPurchaseOrder = record
    POId: Integer;
    ProviderName: string;
    OrderDate: TDateTime;
    ExpectedDelivery: TDateTime;
    TotalAmount: Currency;
    Status: string;
    ItemCount: Integer;
  end;

  TSupplyPriceHistory = record
    ItemName: string;
    ProviderName: string;
    OldPrice: Currency;
    NewPrice: Currency;
    ChangeDate: TDateTime;
    ChangePct: Double;
  end;

  TProviderSuppliesBL = class
  public
    function GetSupplies: TArray<TSupplyItem>;
    function GetPurchaseOrders: TArray<TPurchaseOrder>;
    function GetPriceHistory: TArray<TSupplyPriceHistory>;
  end;

const
  SQL_GET_SUPPLIES =
    'SELECT s.SupplyId, s.ProviderId, p.CompanyName AS ProviderName, ' +
    '       s.ItemName, s.Category, s.UnitCost, s.MinOrderQty, ' +
    '       s.LeadTimeDays, s.InStock ' +
    'FROM Supplies s ' +
    'INNER JOIN Providers p ON p.ProviderId = s.ProviderId ' +
    'ORDER BY s.Category, s.ItemName';

  SQL_GET_PURCHASE_ORDERS =
    'SELECT po.POId, p.CompanyName AS ProviderName, po.OrderDate, ' +
    '       po.ExpectedDelivery, po.TotalAmount, po.Status, ' +
    '       (SELECT COUNT(*) FROM PurchaseOrderItems poi WHERE poi.POId = po.POId) AS ItemCount ' +
    'FROM PurchaseOrders po ' +
    'INNER JOIN Providers p ON p.ProviderId = po.ProviderId ' +
    'ORDER BY po.OrderDate DESC';

  SQL_GET_PRICE_HISTORY =
    'SELECT s.ItemName, p.CompanyName AS ProviderName, ' +
    '       ph.OldPrice, ph.NewPrice, ph.ChangeDate, ' +
    '       ((ph.NewPrice - ph.OldPrice) / ph.OldPrice * 100) AS ChangePct ' +
    'FROM PriceHistory ph ' +
    'INNER JOIN Supplies s ON s.SupplyId = ph.SupplyId ' +
    'INNER JOIN Providers p ON p.ProviderId = s.ProviderId ' +
    'ORDER BY ph.ChangeDate DESC';

  SQL_CREATE_PURCHASE_ORDER =
    'INSERT INTO PurchaseOrders (ProviderId, OrderDate, ExpectedDelivery, ' +
    '                            TotalAmount, Status) ' +
    'VALUES (:ProviderId, :OrderDate, :ExpectedDelivery, ' +
    '        :TotalAmount, :Status)';

implementation

{ TProviderSuppliesBL }

function TProviderSuppliesBL.GetSupplies: TArray<TSupplyItem>;

  procedure AddSupply(var Arr: TArray<TSupplyItem>; AId, AProviderId: Integer;
    const AProvider, AName, ACategory: string; ACost: Currency;
    AMinQty, ALeadDays: Integer; AInStock: Boolean);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].SupplyId := AId;
    Arr[Idx].ProviderId := AProviderId;
    Arr[Idx].ProviderName := AProvider;
    Arr[Idx].ItemName := AName;
    Arr[Idx].Category := ACategory;
    Arr[Idx].UnitCost := ACost;
    Arr[Idx].MinOrderQty := AMinQty;
    Arr[Idx].LeadTimeDays := ALeadDays;
    Arr[Idx].InStock := AInStock;
  end;

begin
  Result := nil;
  AddSupply(Result, 1,  1, 'Fresh Farms Co',       'Tomatoes',         'Produce',   3.20, 10, 1, True);
  AddSupply(Result, 2,  1, 'Fresh Farms Co',       'Garlic',           'Produce',   6.00,  5, 1, True);
  AddSupply(Result, 3,  1, 'Fresh Farms Co',       'Lemons',           'Produce',   4.00,  5, 1, True);
  AddSupply(Result, 4,  2, 'Ocean Direct',          'Salmon Fillet',    'Seafood',  22.00,  5, 2, True);
  AddSupply(Result, 5,  2, 'Ocean Direct',          'Shrimp',           'Seafood',  28.00,  3, 2, True);
  AddSupply(Result, 6,  3, 'Valley Meats',          'Chicken Breast',   'Meat',      9.50, 10, 2, True);
  AddSupply(Result, 7,  4, 'Dairy Delights',        'Mozzarella',       'Dairy',    12.00,  5, 1, True);
  AddSupply(Result, 8,  4, 'Dairy Delights',        'Butter',           'Dairy',     7.50,  5, 1, False);
  AddSupply(Result, 9,  4, 'Dairy Delights',        'Heavy Cream',      'Dairy',     5.50, 10, 1, True);
  AddSupply(Result, 10, 5, 'Golden Grain Supply',   'Flour',            'Dry Goods', 2.80, 25, 3, True);
  AddSupply(Result, 11, 5, 'Golden Grain Supply',   'Rice',             'Dry Goods', 4.50, 25, 3, True);
  AddSupply(Result, 12, 6, 'Mediterranean Imports',  'Olive Oil',        'Oils',     18.50, 10, 5, True);
end;

function TProviderSuppliesBL.GetPurchaseOrders: TArray<TPurchaseOrder>;

  procedure AddPO(var Arr: TArray<TPurchaseOrder>; AId: Integer;
    const AProvider: string; AOrderDate, ADelivery: TDateTime;
    ATotal: Currency; const AStatus: string; AItems: Integer);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].POId := AId;
    Arr[Idx].ProviderName := AProvider;
    Arr[Idx].OrderDate := AOrderDate;
    Arr[Idx].ExpectedDelivery := ADelivery;
    Arr[Idx].TotalAmount := ATotal;
    Arr[Idx].Status := AStatus;
    Arr[Idx].ItemCount := AItems;
  end;

begin
  Result := nil;
  AddPO(Result, 1045, 'Fresh Farms Co',       EncodeDate(2026, 7, 28), EncodeDate(2026, 7, 29),  320.00, 'Delivered',  3);
  AddPO(Result, 1046, 'Ocean Direct',          EncodeDate(2026, 7, 30), EncodeDate(2026, 8, 1),   610.00, 'Delivered',  2);
  AddPO(Result, 1047, 'Valley Meats',          EncodeDate(2026, 8, 1),  EncodeDate(2026, 8, 3),   475.00, 'In Transit', 1);
  AddPO(Result, 1048, 'Dairy Delights',        EncodeDate(2026, 8, 2),  EncodeDate(2026, 8, 3),   285.00, 'Pending',    3);
  AddPO(Result, 1049, 'Golden Grain Supply',   EncodeDate(2026, 8, 3),  EncodeDate(2026, 8, 6),   245.00, 'Pending',    2);
  AddPO(Result, 1050, 'Mediterranean Imports',  EncodeDate(2026, 7, 15), EncodeDate(2026, 7, 20),  370.00, 'Cancelled',  1);
end;

function TProviderSuppliesBL.GetPriceHistory: TArray<TSupplyPriceHistory>;

  procedure AddHistory(var Arr: TArray<TSupplyPriceHistory>;
    const AItem, AProvider: string; AOld, ANew: Currency;
    ADate: TDateTime; APct: Double);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].ItemName := AItem;
    Arr[Idx].ProviderName := AProvider;
    Arr[Idx].OldPrice := AOld;
    Arr[Idx].NewPrice := ANew;
    Arr[Idx].ChangeDate := ADate;
    Arr[Idx].ChangePct := APct;
  end;

begin
  Result := nil;
  AddHistory(Result, 'Salmon Fillet',   'Ocean Direct',          20.00, 22.00, EncodeDate(2026, 7, 1),   10.0);
  AddHistory(Result, 'Olive Oil',       'Mediterranean Imports',  17.00, 18.50, EncodeDate(2026, 6, 15),   8.82);
  AddHistory(Result, 'Chicken Breast',  'Valley Meats',           9.00,  9.50, EncodeDate(2026, 6, 1),    5.56);
  AddHistory(Result, 'Mozzarella',      'Dairy Delights',        11.50, 12.00, EncodeDate(2026, 5, 20),   4.35);
  AddHistory(Result, 'Flour',           'Golden Grain Supply',    3.00,  2.80, EncodeDate(2026, 5, 1),   -6.67);
  AddHistory(Result, 'Tomatoes',        'Fresh Farms Co',         3.50,  3.20, EncodeDate(2026, 4, 15),  -8.57);
  AddHistory(Result, 'Heavy Cream',     'Dairy Delights',         5.00,  5.50, EncodeDate(2026, 4, 1),   10.0);
  AddHistory(Result, 'Shrimp',          'Ocean Direct',          26.00, 28.00, EncodeDate(2026, 3, 15),   7.69);
end;

end.

File: ProvidersSupplies.Core/Application/Services/ProvidersSupplies/uProviderSuppliesBL.cs
using Microsoft.Extensions.Logging;
using ProvidersSupplies.Core.Application.DTOs.Core;

namespace ProvidersSupplies.Core.Application.Services.ProvidersSupplies;

public class uProviderSuppliesBL : IuProviderSuppliesBL
{
    private readonly ILogger<uProviderSuppliesBL> _logger;

    public uProviderSuppliesBL(ILogger<uProviderSuppliesBL> logger)
    {
        _logger = logger;
    }

    /// <summary>
    /// Ported from TProviderSuppliesBL.GetSupplies — returns the full supply catalog
    /// with provider information, ordered by Category then ItemName.
    /// Legacy source: hardcoded seed data representing what SQL_GET_SUPPLIES would return.
    /// </summary>
    public TSupplyItem[] GetSupplies()
    {
        _logger.LogDebug("GetSupplies called");

        return new TSupplyItem[]
        {
            new TSupplyItem { SupplyId = 1,  ProviderId = 1, ProviderName = "Fresh Farms Co",       ItemName = "Tomatoes",       Category = "Produce",   UnitCost = 3.20m,  MinOrderQty = 10, LeadTimeDays = 1, InStock = true },
            new TSupplyItem { SupplyId = 2,  ProviderId = 1, ProviderName = "Fresh Farms Co",       ItemName = "Garlic",         Category = "Produce",   UnitCost = 6.00m,  MinOrderQty = 5,  LeadTimeDays = 1, InStock = true },
            new TSupplyItem { SupplyId = 3,  ProviderId = 1, ProviderName = "Fresh Farms Co",       ItemName = "Lemons",         Category = "Produce",   UnitCost = 4.00m,  MinOrderQty = 5,  LeadTimeDays = 1, InStock = true },
            new TSupplyItem { SupplyId = 4,  ProviderId = 2, ProviderName = "Ocean Direct",         ItemName = "Salmon Fillet",  Category = "Seafood",   UnitCost = 22.00m, MinOrderQty = 5,  LeadTimeDays = 2, InStock = true },
            new TSupplyItem { SupplyId = 5,  ProviderId = 2, ProviderName = "Ocean Direct",         ItemName = "Shrimp",         Category = "Seafood",   UnitCost = 28.00m, MinOrderQty = 3,  LeadTimeDays = 2, InStock = true },
            new TSupplyItem { SupplyId = 6,  ProviderId = 3, ProviderName = "Valley Meats",         ItemName = "Chicken Breast", Category = "Meat",      UnitCost = 9.50m,  MinOrderQty = 10, LeadTimeDays = 2, InStock = true },
            new TSupplyItem { SupplyId = 7,  ProviderId = 4, ProviderName = "Dairy Delights",       ItemName = "Mozzarella",     Category = "Dairy",     UnitCost = 12.00m, MinOrderQty = 5,  LeadTimeDays = 1, InStock = true },
            new TSupplyItem { SupplyId = 8,  ProviderId = 4, ProviderName = "Dairy Delights",       ItemName = "Butter",         Category = "Dairy",     UnitCost = 7.50m,  MinOrderQty = 5,  LeadTimeDays = 1, InStock = false },
            new TSupplyItem { SupplyId = 9,  ProviderId = 4, ProviderName = "Dairy Delights",       ItemName = "Heavy Cream",    Category = "Dairy",     UnitCost = 5.50m,  MinOrderQty = 10, LeadTimeDays = 1, InStock = true },
            new TSupplyItem { SupplyId = 10, ProviderId = 5, ProviderName = "Golden Grain Supply",  ItemName = "Flour",          Category = "Dry Goods", UnitCost = 2.80m,  MinOrderQty = 25, LeadTimeDays = 3, InStock = true },
            new TSupplyItem { SupplyId = 11, ProviderId = 5, ProviderName = "Golden Grain Supply",  ItemName = "Rice",           Category = "Dry Goods", UnitCost = 4.50m,  MinOrderQty = 25, LeadTimeDays = 3, InStock = true },
            new TSupplyItem { SupplyId = 12, ProviderId = 6, ProviderName = "Mediterranean Imports", ItemName = "Olive Oil",     Category = "Oils",      UnitCost = 18.50m, MinOrderQty = 10, LeadTimeDays = 5, InStock = true }
        };
    }

    /// <summary>
    /// Ported from TProviderSuppliesBL.GetPurchaseOrders — returns purchase orders
    /// ordered by OrderDate descending.
    /// Legacy source: hardcoded seed data representing what SQL_GET_PURCHASE_ORDERS would return.
    /// </summary>
    public TPurchaseOrder[] GetPurchaseOrders()
    {
        _logger.LogDebug("GetPurchaseOrders called");

        return new TPurchaseOrder[]
        {
            new TPurchaseOrder { POId = 1045, ProviderName = "Fresh Farms Co",       OrderDate = new DateTime(2026, 7, 28, 0, 0, 0, DateTimeKind.Utc), ExpectedDelivery = new DateTime(2026, 7, 29, 0, 0, 0, DateTimeKind.Utc), TotalAmount = 320.00m, Status = "Delivered",  ItemCount = 3 },
            new TPurchaseOrder { POId = 1046, ProviderName = "Ocean Direct",         OrderDate = new DateTime(2026, 7, 30, 0, 0, 0, DateTimeKind.Utc), ExpectedDelivery = new DateTime(2026, 8, 1,  0, 0, 0, DateTimeKind.Utc), TotalAmount = 610.00m, Status = "Delivered",  ItemCount = 2 },
            new TPurchaseOrder { POId = 1047, ProviderName = "Valley Meats",         OrderDate = new DateTime(2026, 8, 1,  0, 0, 0, DateTimeKind.Utc), ExpectedDelivery = new DateTime(2026, 8, 3,  0, 0, 0, DateTimeKind.Utc), TotalAmount = 475.00m, Status = "In Transit", ItemCount = 1 },
            new TPurchaseOrder { POId = 1048, ProviderName = "Dairy Delights",       OrderDate = new DateTime(2026, 8, 2,  0, 0, 0, DateTimeKind.Utc), ExpectedDelivery = new DateTime(2026, 8, 3,  0, 0, 0, DateTimeKind.Utc), TotalAmount = 285.00m, Status = "Pending",    ItemCount = 3 },
            new TPurchaseOrder { POId = 1049, ProviderName = "Golden Grain Supply",  OrderDate = new DateTime(2026, 8, 3,  0, 0, 0, DateTimeKind.Utc), ExpectedDelivery = new DateTime(2026, 8, 6,  0, 0, 0, DateTimeKind.Utc), TotalAmount = 245.00m, Status = "Pending",    ItemCount = 2 },
            new TPurchaseOrder { POId = 1050, ProviderName = "Mediterranean Imports", OrderDate = new DateTime(2026, 7, 15, 0, 0, 0, DateTimeKind.Utc), ExpectedDelivery = new DateTime(2026, 7, 20, 0, 0, 0, DateTimeKind.Utc), TotalAmount = 370.00m, Status = "Cancelled",  ItemCount = 1 }
        };
    }

    /// <summary>
    /// Ported from TProviderSuppliesBL.GetPriceHistory — returns supply price change history
    /// ordered by ChangeDate descending.
    /// Legacy source: hardcoded seed data representing what SQL_GET_PRICE_HISTORY would return.
    /// ChangePct is computed as ((NewPrice - OldPrice) / OldPrice * 100).
    /// </summary>
    public TSupplyPriceHistory[] GetPriceHistory()
    {
        _logger.LogDebug("GetPriceHistory called");

        return new TSupplyPriceHistory[]
        {
            new TSupplyPriceHistory { ItemName = "Salmon Fillet",  ProviderName = "Ocean Direct",         OldPrice = 20.00m, NewPrice = 22.00m, ChangeDate = new DateTime(2026, 7, 1,  0, 0, 0, DateTimeKind.Utc), ChangePct = 10.0 },
            new TSupplyPriceHistory { ItemName = "Olive Oil",      ProviderName = "Mediterranean Imports", OldPrice = 17.00m, NewPrice = 18.50m, ChangeDate = new DateTime(2026, 6, 15, 0, 0, 0, DateTimeKind.Utc), ChangePct = 8.82 },
            new TSupplyPriceHistory { ItemName = "Chicken Breast", ProviderName = "Valley Meats",         OldPrice = 9.00m,  NewPrice = 9.50m,  ChangeDate = new DateTime(2026, 6, 1,  0, 0, 0, DateTimeKind.Utc), ChangePct = 5.56 },
            new TSupplyPriceHistory { ItemName = "Mozzarella",     ProviderName = "Dairy Delights",       OldPrice = 11.50m, NewPrice = 12.00m, ChangeDate = new DateTime(2026, 5, 20, 0, 0, 0, DateTimeKind.Utc), ChangePct = 4.35 },
            new TSupplyPriceHistory { ItemName = "Flour",          ProviderName = "Golden Grain Supply",  OldPrice = 3.00m,  NewPrice = 2.80m,  ChangeDate = new DateTime(2026, 5, 1,  0, 0, 0, DateTimeKind.Utc), ChangePct = -6.67 },
            new TSupplyPriceHistory { ItemName = "Tomatoes",       ProviderName = "Fresh Farms Co",       OldPrice = 3.50m,  NewPrice = 3.20m,  ChangeDate = new DateTime(2026, 4, 15, 0, 0, 0, DateTimeKind.Utc), ChangePct = -8.57 },
            new TSupplyPriceHistory { ItemName = "Heavy Cream",    ProviderName = "Dairy Delights",       OldPrice = 5.00m,  NewPrice = 5.50m,  ChangeDate = new DateTime(2026, 4, 1,  0, 0, 0, DateTimeKind.Utc), ChangePct = 10.0 },
            new TSupplyPriceHistory { ItemName = "Shrimp",         ProviderName = "Ocean Direct",         OldPrice = 26.00m, NewPrice = 28.00m, ChangeDate = new DateTime(2026, 3, 15, 0, 0, 0, DateTimeKind.Utc), ChangePct = 7.69 }
        };
    }

    /// <summary>
    /// Async wrapper for GetSupplies.
    /// </summary>
    public async Task<List<TSupplyItem>> GetSuppliesAsync(CancellationToken ct = default)
    {
        return await Task.FromResult(new List<TSupplyItem>(GetSupplies())).ConfigureAwait(false);
    }

    /// <summary>
    /// Async wrapper for GetPurchaseOrders.
    /// </summary>
    public async Task<List<TPurchaseOrder>> GetPurchaseOrdersAsync(CancellationToken ct = default)
    {
        return await Task.FromResult(new List<TPurchaseOrder>(GetPurchaseOrders())).ConfigureAwait(false);
    }

    /// <summary>
    /// Async wrapper for GetPriceHistory.
    /// </summary>
    public async Task<List<TSupplyPriceHistory>> GetPriceHistoryAsync(CancellationToken ct = default)
    {
        return await Task.FromResult(new List<TSupplyPriceHistory>(GetPriceHistory())).ConfigureAwait(false);
    }
}
